presto-orc/src/main/java/com/facebook/presto/orc/stream/ByteOutputStream.java

```diff
@@ -0,0 +1,184 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.orc.stream;
+
+import com.facebook.presto.orc.OrcOutputBuffer;
+import com.facebook.presto.orc.checkpoint.ByteStreamCheckpoint;
+import com.facebook.presto.orc.metadata.CompressionKind;
+import com.facebook.presto.orc.metadata.Stream;
+import com.google.common.collect.ImmutableList;
+import io.airlift.slice.SizeOf;
+import io.airlift.slice.SliceOutput;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+
+import static com.facebook.presto.orc.metadata.Stream.StreamKind.DATA;
+import static com.google.common.base.Preconditions.checkState;
+
+public class ByteOutputStream
+        implements ValueOutputStream<ByteStreamCheckpoint>
+{
+    // todo use OrcStreamUtils
+    private static final int MIN_REPEAT_SIZE = 3;
+    private static final int UNMATCHABLE_VALUE = Integer.MAX_VALUE;
+
+    private final OrcOutputBuffer buffer;
+    private final List<ByteStreamCheckpoint> checkpoints = new ArrayList<>();
+
+    private final byte[] sequenceBuffer = new byte[128];
+    private int size;
+
+    private int runCount;
+    private int lastValue = UNMATCHABLE_VALUE;
+
+    private boolean closed;
+
+    public ByteOutputStream(CompressionKind compression, int bufferSize)
+    {
+        this(new OrcOutputBuffer(compression, bufferSize));
+    }
+
+    public ByteOutputStream(OrcOutputBuffer buffer)
+    {
+        this.buffer = buffer;
+    }
+
+    @Override
+    public Class<ByteStreamCheckpoint> getCheckpointType()
+    {
+        return ByteStreamCheckpoint.class;
+    }
+
+    public void writeByte(byte value)
+    {
+        checkState(!closed);
+
+        // flush if buffer is full
+        if (size == sequenceBuffer.length) {
+            flushSequence();
+        }
+
+        // update run count
+        if (value == lastValue) {
+            runCount++;
+        }
+        else {
+            // run ended, so flush
+            if (runCount >= MIN_REPEAT_SIZE) {
+                flushSequence();
+            }
+            runCount = 1;
+        }
+
+        // buffer value
+        sequenceBuffer[size] = value;
+        size++;
+
```

Comment:
As I added at the end of that comment. It's up to you. I probably should have put that at the beginning.